tests/crypto_aesctr/main.c

```diff
@@ -0,0 +1,448 @@
+#include <sys/time.h>
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "cpusupport.h"
+#include "crypto_aes.h"
+#include "crypto_aesctr.h"
+#include "getopt.h"
+#include "hexify.h"
+#include "insecure_memzero.h"
+#include "monoclock.h"
+#include "warnp.h"
+
+#define MAX_PLAINTEXT_LENGTH 32
+
+#define LARGE_BUFSIZE 65536
+#define MAX_CHUNK 256
+
+/* Forward declaration. */
+struct crypto_aes_key;
+
+struct testcase {
+	const char * keytext_hex;
+	const char * plaintext_str;
+	const char * ciphertext_hex;
+};
+
+/* Test vectors. */
+static const struct testcase tests[] = {
+	/* 128-bit AES-CTR. */
+	{"000102030405060708090a0b0c0d0e0f",
+	    " ",
+	    "e6"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "A",
+	    "87"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "AAAA",
+	    "87e07a76"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "AB",
+	    "87e3"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "hello",
+	    "aec4575be8"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "hello world",
+	    "aec4575be8af2ced1d23e5"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "This is 16 chars",
+	    "92c95244a7e628a25e79a101c9a9aa0a"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "Ceci n'est pas 24 chars.",
+	    "85c4585ea7e17ce71c3ba112c0bbf84b476670fdf4b2c730"},
+	{"000102030405060708090a0b0c0d0e0f",
+	    "This block is exactly 32 chars!!",
+	    "92c95244a7ed37ed0c24a10bd2e8bd01122567f9ece0872c6918d58217870c2b"},
+	/* 256-bit AES-CTR. */
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    " ",
+	    "d2"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "A",
+	    "b3"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "AAAA",
+	    "b3d141f7"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "AB",
+	    "b3d2"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "hello",
+	    "9af56cda45"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "hello world",
+	    "9af56cda4569e8bfdb9ffe"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "This is 16 chars",
+	    "a6f869c50a20ecf098c5ba09b54f05f3"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "Ceci n'est pas 24 chars.",
+	    "b1f563df0a27b8b5da87ba1abc5d57b2c47d15c62bcbeccb"},
+	{"000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f",
+	    "This block is exactly 32 chars!!",
+	    "a6f869c50a2bf3bfca98ba03ae0e12f8913e02c23399acd78695f3503ab1171c"}
+};
+
+/* Largest buffer must be first.  1028 is what's used in spiped. */
+static const size_t perfsizes[] = {16384, 8192, 4096, 2048, 1028, 1024, 512, 256,
+    128, 64, 32, 16};
```

Comment:
The first idea for benchmarking AES-CTR was "do the same thing as the crc32 test".  But are all these sizes relevant?  spiped only uses 1028-byte buffers.  scrypt uses variable size[1], but if somebody is only trying to encrypt 16 bytes then aesctr isn't going to be the bottleneck.  Tarsnap is variable as well.

[1] for buffers, it's the entire thing at once.  For files, it's up to 65536 at once.

FWIW, the drop-off on my desktop machine (of the un-optimized code) is:

```
$ ./test_crypto_aesctr -t
Performance test of AES-CTR using hardware AESNI.
... in 32896 blocks of size 16384:	2.003606 s	269.0 MB/s
... in 65792 blocks of size 8192:	2.002147 s	269.2 MB/s
... in 131584 blocks of size 4096:	2.009137 s	268.3 MB/s
... in 263168 blocks of size 2048:	2.003082 s	269.1 MB/s
... in 524288 blocks of size 1028:	2.025053 s	266.2 MB/s
... in 526336 blocks of size 1024:	2.005457 s	268.8 MB/s
... in 1052672 blocks of size 512:	2.015918 s	267.4 MB/s
... in 2105344 blocks of size 256:	2.041611 s	264.0 MB/s
... in 4210688 blocks of size 128:	2.094933 s	257.3 MB/s
... in 8421376 blocks of size 64:	2.205024 s	244.4 MB/s
... in 16842752 blocks of size 32:	2.419742 s	222.7 MB/s
... in 33685504 blocks of size 16:	2.854426 s	188.8 MB/s
Overall time and speed:	25.68 s	251.9 MB/s
```

Reply:
Maybe make it sizes 1--1024?  That seems like the interesting range.

Reply:
Including 1028 for testing exactly how it'll be used in spiped, or just end at 1024?

Reply:
Just end at 1024.  The speed for 1028 isn't much different and I don't really want libcperciva polluted by spiped-specific details.

Reply:
Hmm, are you sure you want to go all the way down to 1?  Like, `crypto_aesctr_buf(key_exp, 0, buf, buf, 1);`?  It's atrociously slow, but surely nobody would be calling it on a 1-byte buffer for any kind of time-sensitive task.

```
$ ./test_crypto_aesctr -t
Performance test of AES-CTR using hardware AESNI.
... in 526336 blocks of size 1024:	2.010606 s	268.1 MB/s
... in 1052672 blocks of size 512:	2.022461 s	266.5 MB/s
... in 2105344 blocks of size 256:	2.049952 s	262.9 MB/s
... in 4210688 blocks of size 128:	2.095254 s	257.2 MB/s
... in 8421376 blocks of size 64:	2.204425 s	244.5 MB/s
... in 16842752 blocks of size 32:	2.420215 s	222.7 MB/s
... in 33685504 blocks of size 16:	2.853046 s	188.9 MB/s
... in 67371008 blocks of size 8:	5.026268 s	107.2 MB/s
... in 134742016 blocks of size 4:	9.405395 s	57.3 MB/s
... in 269484032 blocks of size 2:	18.148134 s	29.7 MB/s
... in 538968064 blocks of size 1:	35.610145 s	15.1 MB/s
Overall time and speed:	83.85 s	70.7 MB/s
```

If you really want this, ok, but in that case I think we should omit the "Overall time and speed", because for any serious use we'd surely be using bigger buffer sizes.
(I'm also going to play with a smaller total byte size.)

Reply:
Yes, "overall time and speed" is a red herring so I'd drop that.

Yes, nobody should be encrypting one byte at a time; in a way, that's why I think it's useful to have the benchmark show how terrible that is. ;-)

I'd drop the total number of bytes, maybe down to 32 MB.

Reply:
Ok.
```
$ ./test_crypto_aesctr -t
Performance test of AES-CTR using hardware AESNI.
... in 32768 blocks of size 1024:	0.125179 s	268.1 MB/s
... in 65536 blocks of size 512:	0.125577 s	267.2 MB/s
... in 131072 blocks of size 256:	0.127058 s	264.1 MB/s
... in 262144 blocks of size 128:	0.130472 s	257.2 MB/s
... in 524288 blocks of size 64:	0.137340 s	244.3 MB/s
... in 1048576 blocks of size 32:	0.150582 s	222.8 MB/s
... in 2097152 blocks of size 16:	0.184997 s	181.4 MB/s
... in 4194304 blocks of size 8:	0.312804 s	107.3 MB/s
... in 8388608 blocks of size 4:	0.583361 s	57.5 MB/s
... in 16777216 blocks of size 2:	1.129428 s	29.7 MB/s
... in 33554432 blocks of size 1:	2.216826 s	15.1 MB/s
$
```